File: curator/thetadata-backfill-rs/src/models/equity.rs
//! Equity data models.

use chrono::{DateTime, NaiveDate, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// Equity OHLC data
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct EquityOhlc {
    #[serde(skip_deserializing)]
    pub id: Option<i64>,
    pub symbol: String,
    pub timestamp: DateTime<Utc>,
    pub open: Decimal,
    pub high: Decimal,
    pub low: Decimal,
    pub close: Decimal,
    pub volume: i64,
    pub count: Option<i64>,
}

/// Equity trade data
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct EquityTrade {
    #[serde(skip_deserializing)]
    pub id: Option<i64>,
    pub symbol: String,
    pub timestamp: DateTime<Utc>,
    pub sequence: i64,
    pub size: i64,
    pub price: Decimal,
    pub condition: i32,
    pub exchange: Option<i32>,
    pub ext_condition1: Option<i32>,
    pub ext_condition2: Option<i32>,
    pub ext_condition3: Option<i32>,
    pub ext_condition4: Option<i32>,
}

/// Equity quote (NBBO) data
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct EquityQuote {
    #[serde(skip_deserializing)]
    pub id: Option<i64>,
    pub symbol: String,
    pub timestamp: DateTime<Utc>,
    pub bid: Decimal,
    pub bid_size: i64,
    pub bid_exchange: Option<i32>,
    pub bid_condition: Option<i32>,
    pub ask: Decimal,
    pub ask_size: i64,
    pub ask_exchange: Option<i32>,
    pub ask_condition: Option<i32>,
}

/// Equity end-of-day report
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct EquityEod {
    #[serde(skip_deserializing)]
    pub id: Option<i64>,
    pub symbol: String,
    pub trade_date: NaiveDate,
    pub report_created: DateTime<Utc>,
    pub last_trade: Option<DateTime<Utc>>,
    pub open: Decimal,
    pub high: Decimal,
    pub low: Decimal,
    pub close: Decimal,
    pub volume: i64,
    pub count: i64,
    pub bid: Option<Decimal>,
    pub bid_size: Option<i64>,
    pub ask: Option<Decimal>,
    pub ask_size: Option<i64>,
}

/// Equity OHLC API response
#[derive(Debug, Clone, Deserialize)]
pub struct EquityOhlcResponse {
    pub timestamp: DateTime<Utc>,
    pub symbol: Option<String>,
    pub open: Decimal,
    pub high: Decimal,
    pub low: Decimal,
    pub close: Decimal,
    pub volume: i64,
    pub count: Option<i64>,
}

impl EquityOhlcResponse {
    pub fn to_record(self, symbol: &str) -> EquityOhlc {
        EquityOhlc {
            id: None,
            symbol: self.symbol.unwrap_or_else(|| symbol.to_string()),
            timestamp: self.timestamp,
            open: self.open,
            high: self.high,
            low: self.low,
            close: self.close,
            volume: self.volume,
            count: self.count,
        }
    }
}

/// Equity trade API response
#[derive(Debug, Clone, Deserialize)]
pub struct EquityTradeResponse {
    pub timestamp: DateTime<Utc>,
    pub symbol: Option<String>,
    pub sequence: i64,
    pub size: i64,
    pub price: Decimal,
    pub condition: i32,
    pub exchange: Option<i32>,
    pub ext_condition1: Option<i32>,
    pub ext_condition2: Option<i32>,
    pub ext_condition3: Option<i32>,
    pub ext_condition4: Option<i32>,
}

impl EquityTradeResponse {
    pub fn to_record(self, symbol: &str) -> EquityTrade {
        EquityTrade {
            id: None,
            symbol: self.symbol.unwrap_or_else(|| symbol.to_string()),
            timestamp: self.timestamp,
            sequence: self.sequence,
            size: self.size,
            price: self.price,
            condition: self.condition,
            exchange: self.exchange,
            ext_condition1: self.ext_condition1,
            ext_condition2: self.ext_condition2,
            ext_condition3: self.ext_condition3,
            ext_condition4: self.ext_condition4,
        }
    }
}

/// Equity quote API response
#[derive(Debug, Clone, Deserialize)]
pub struct EquityQuoteResponse {
    pub timestamp: DateTime<Utc>,
    pub symbol: Option<String>,
    pub bid: Decimal,
    pub bid_size: i64,
    pub bid_exchange: Option<i32>,
    pub bid_condition: Option<i32>,
    pub ask: Decimal,
    pub ask_size: i64,
    pub ask_exchange: Option<i32>,
    pub ask_condition: Option<i32>,
}

impl EquityQuoteResponse {
    pub fn to_record(self, symbol: &str) -> EquityQuote {
        EquityQuote {
            id: None,
            symbol: self.symbol.unwrap_or_else(|| symbol.to_string()),
            timestamp: self.timestamp,
            bid: self.bid,
            bid_size: self.bid_size,
            bid_exchange: self.bid_exchange,
            bid_condition: self.bid_condition,
            ask: self.ask,
            ask_size: self.ask_size,
            ask_exchange: self.ask_exchange,
            ask_condition: self.ask_condition,
        }
    }
}

/// Equity EOD API response
#[derive(Debug, Clone, Deserialize)]
pub struct EquityEodResponse {
    pub created: DateTime<Utc>,
    pub last_trade: Option<DateTime<Utc>>,
    pub symbol: Option<String>,
    pub open: Decimal,
    pub high: Decimal,
    pub low: Decimal,
    pub close: Decimal,
    pub volume: i64,
    #[serde(default)]
    pub count: i64,
    pub bid: Option<Decimal>,
    pub bid_size: Option<i64>,
    pub ask: Option<Decimal>,
    pub ask_size: Option<i64>,
}

impl EquityEodResponse {
    pub fn to_record(self, symbol: &str) -> EquityEod {
        EquityEod {
            id: None,
            symbol: self.symbol.unwrap_or_else(|| symbol.to_string()),
            trade_date: self.created.date_naive(),
            report_created: self.created,
            last_trade: self.last_trade,
            open: self.open,
            high: self.high,
            low: self.low,
            close: self.close,
            volume: self.volume,
            count: self.count,
            bid: self.bid,
            bid_size: self.bid_size,
            ask: self.ask,
            ask_size: self.ask_size,
        }
    }
}

File: curator/thetadata/thetadata_backfill/downloaders/equity.py
"""Equity data downloader."""

from datetime import date
from typing import Any

from thetadata_backfill.downloaders.base import BaseDownloader
from thetadata_backfill.models.equity import EquityEOD, EquityOHLC, EquityQuote, EquityTrade
from thetadata_backfill.schemas.equity import (
    EquityEODResponse,
    EquityOHLCResponse,
    EquityQuoteResponse,
    EquityTradeResponse,
)


class EquityOHLCDownloader(BaseDownloader[EquityOHLCResponse, EquityOHLC]):
    """Downloader for equity OHLC data."""
    
    data_type = "ohlc"
    schema_class = EquityOHLCResponse
    model_class = EquityOHLC
    
    def get_list_endpoint(self) -> str:
        return "stock/list/symbols"
    
    def get_dates_endpoint(self, request_type: str = "ohlc") -> str:
        return f"stock/list/dates/{request_type}"
    
    def get_history_endpoint(self, request_type: str) -> str:
        return f"stock/history/{request_type}"
    
    def build_request_params(
        self,
        symbol: str,
        target_date: date,
        request_type: str,
        interval: str = "1m",
        **kwargs: Any,
    ) -> dict[str, Any]:
        return {
            "symbol": symbol,
            "date": target_date.strftime("%Y%m%d"),
            "interval": interval,
        }
    
    def parse_response(
        self,
        data: dict[str, Any],
        symbol: str,
    ) -> list[EquityOHLCResponse]:
        return [
            EquityOHLCResponse(**{**item, "symbol": item.get("symbol", symbol)})
            for item in data["response"]
        ]
    
    def schema_to_record(self, schema: EquityOHLCResponse, symbol: str) -> dict[str, Any]:
        return {
            "symbol": schema.symbol,
            "timestamp": schema.timestamp,
            "open": schema.open,
            "high": schema.high,
            "low": schema.low,
            "close": schema.close,
            "volume": schema.volume,
            "count": schema.count,
        }
    
    def get_conflict_columns(self) -> list[str]:
        return ["symbol", "timestamp"]
    
    def get_update_columns(self) -> list[str]:
        return ["open", "high", "low", "close", "volume", "count"]


class EquityTradeDownloader(BaseDownloader[EquityTradeResponse, EquityTrade]):
    """Downloader for equity trade data."""
    
    data_type = "trade"
    schema_class = EquityTradeResponse
    model_class = EquityTrade
    
    def get_list_endpoint(self) -> str:
        return "stock/list/symbols"
    
    def get_dates_endpoint(self, request_type: str = "trade") -> str:
        return f"stock/list/dates/{request_type}"
    
    def get_history_endpoint(self, request_type: str) -> str:
        return f"stock/history/{request_type}"
    
    def build_request_params(
        self,
        symbol: str,
        target_date: date,
        request_type: str,
        **kwargs: Any,
    ) -> dict[str, Any]:
        return {
            "symbol": symbol,
            "date": target_date.strftime("%Y%m%d"),
        }
    
    def parse_response(
        self,
        data: dict[str, Any],
        symbol: str,
    ) -> list[EquityTradeResponse]:
        return [
            EquityTradeResponse(**{**item, "symbol": item.get("symbol", symbol)})
            for item in data["response"]
        ]
    
    def schema_to_record(self, schema: EquityTradeResponse, symbol: str) -> dict[str, Any]:
        return {
            "symbol": symbol if schema.symbol is None else schema.symbol,
            "timestamp": schema.timestamp,
            "sequence": schema.sequence,
            "size": schema.size,
            "price": schema.price,
            "condition": schema.condition,
            "exchange": schema.exchange,
            "ext_condition1": schema.ext_condition1,
            "ext_condition2": schema.ext_condition2,
            "ext_condition3": schema.ext_condition3,
            "ext_condition4": schema.ext_condition4,
        }
    
    def get_conflict_columns(self) -> list[str]:
        return ["symbol", "sequence"]
    
    def get_update_columns(self) -> list[str]:
        return ["timestamp", "size", "price", "condition"]


class EquityQuoteDownloader(BaseDownloader[EquityQuoteResponse, EquityQuote]):
    """Downloader for equity quote (NBBO) data."""
    
    data_type = "quote"
    schema_class = EquityQuoteResponse
    model_class = EquityQuote
    
    def get_list_endpoint(self) -> str:
        return "stock/list/symbols"
    
    def get_dates_endpoint(self, request_type: str = "quote") -> str:
        return f"stock/list/dates/{request_type}"
    
    def get_history_endpoint(self, request_type: str) -> str:
        return f"stock/history/{request_type}"
    
    def build_request_params(
        self,
        symbol: str,
        target_date: date,
        request_type: str,
        interval: str = "1m",
        **kwargs: Any,
    ) -> dict[str, Any]:
        return {
            "symbol": symbol,
            "date": target_date.strftime("%Y%m%d"),
            "interval": interval,
        }
    
    def parse_response(
        self,
        data: dict[str, Any],
        symbol: str,
    ) -> list[EquityQuoteResponse]:
        return [
            EquityQuoteResponse(**{**item, "symbol": item.get("symbol", symbol)})
            for item in data["response"]
        ]
    
    def schema_to_record(self, schema: EquityQuoteResponse, symbol: str) -> dict[str, Any]:
        return {
            "symbol": symbol if schema.symbol is None else schema.symbol,
            "timestamp": schema.timestamp,
            "bid": schema.bid,
            "bid_size": schema.bid_size,
            "bid_exchange": schema.bid_exchange,
            "bid_condition": schema.bid_condition,
            "ask": schema.ask,
            "ask_size": schema.ask_size,
            "ask_exchange": schema.ask_exchange,
            "ask_condition": schema.ask_condition,
        }
    
    def get_conflict_columns(self) -> list[str]:
        return ["symbol", "timestamp"]
    
    def get_update_columns(self) -> list[str]:
        return ["bid", "bid_size", "ask", "ask_size"]


class EquityEODDownloader(BaseDownloader[EquityEODResponse, EquityEOD]):
    """Downloader for equity end-of-day data."""
    
    data_type = "eod"
    schema_class = EquityEODResponse
    model_class = EquityEOD
    
    def get_list_endpoint(self) -> str:
        return "stock/list/symbols"
    
    def get_dates_endpoint(self, request_type: str = "eod") -> str:
        return f"stock/list/dates/{request_type}"
    
    def get_history_endpoint(self, request_type: str) -> str:
        return "stock/history/eod"
    
    def build_request_params(
        self,
        symbol: str,
        target_date: date,
        request_type: str,
        **kwargs: Any,
    ) -> dict[str, Any]:
        return {
            "symbol": symbol,
            "start_date": target_date.strftime("%Y%m%d"),
            "end_date": target_date.strftime("%Y%m%d"),
        }
    
    def parse_response(
        self,
        data: dict[str, Any],
        symbol: str,
    ) -> list[EquityEODResponse]:
        return [
            EquityEODResponse(**{**item, "symbol": item.get("symbol", symbol)})
            for item in data["response"]
        ]
    
    def schema_to_record(self, schema: EquityEODResponse, symbol: str) -> dict[str, Any]:
        return {
            "symbol": symbol if schema.symbol is None else schema.symbol,
            "trade_date": schema.created.date(),
            "report_created": schema.created,
            "last_trade": schema.last_trade,
            "open": schema.open,
            "high": schema.high,
            "low": schema.low,
            "close": schema.close,
            "volume": schema.volume,
            "count": schema.count,
            "bid": schema.bid,
            "bid_size": schema.bid_size,
            "ask": schema.ask,
            "ask_size": schema.ask_size,
        }
    
    def get_conflict_columns(self) -> list[str]:
        return ["symbol", "trade_date"]
    
    def get_update_columns(self) -> list[str]:
        return ["open", "high", "low", "close", "volume", "count"]
